### PR TITLE
Fix false positive when type casting in IF statement

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/phpstan/phpstan/issues/11674

see https://3v4l.org/NeugB

### CODE REVIEW

File: tests/PHPStan/Analyser/nsrt/narrow-cast.php [hunk 1]
```diff
@@ -63,6 +63,12 @@ function castInt($x, string $s, bool $b) {
         assertType('false', $b);
     }
 
+	if ((int) $s) {
+		assertType('string', $s); // could be *NEVER*
```
Comment (mvorisek) [inline tests/PHPStan/Analyser/nsrt/narrow-cast.php]: reachable for `1` or `01`, `numeric-string&non-falsy-string`?

Comment (staabm) [reply tests/PHPStan/Analyser/nsrt/narrow-cast.php]: thank you

Comment (staabm) [reply tests/PHPStan/Analyser/nsrt/narrow-cast.php]: we reach ELSE for `'0.1'` which is a numeric string - https://3v4l.org/NeugB

Comment (mvorisek) [reply tests/PHPStan/Analyser/nsrt/narrow-cast.php]: and not for `1.5` - https://3v4l.org/VBPn2 - so `string` is best or something like `string~(''|'0')`?

also notice `1x` - https://3v4l.org/PGs9f

Comment (staabm) [reply tests/PHPStan/Analyser/nsrt/narrow-cast.php]: `string~(''|'0')` is `non-falsy-string`. but I think we are in crazy territory land where `string` should be good enough ;)

File: tests/PHPStan/Rules/Comparison/ElseIfConstantConditionRuleTest.php [hunk 2]
```diff
@@ -134,10 +134,6 @@ public function testBug11674(): void
 				'Elseif condition is always false.',
 				28,
 			],
-			[
-				'Elseif condition is always false.',
-				36,
```
Comment (ondrejmirtes) [inline tests/PHPStan/Rules/Comparison/ElseIfConstantConditionRuleTest.php]: Took me a while but I found the bad boy for which the elseif can be false: https://3v4l.org/p1sON

**Comment by ondrejmirtes** [comment]
I think this logic in TypeSpecifier makes more sense. Before the issue we didn't know that we can't compare all the casts to `!= false`. So it makes sense to separate the conditions like this.

We lost precision in a few cases but now it's up to `==` / `!=` resolution logic in TypeSpecifier to make them more precise again. We don't need to do that now of course.

**Comment by ondrejmirtes** [comment]
Thank you.

**Comment by staabm** [comment]
thank you!

**Pull request merged**